src/core/ext/transport/chttp2/transport/hpack_encoder.cc

```diff
@@ -452,10 +452,15 @@ static void add_key(grpc_chttp2_hpack_compressor* c, grpc_mdelem elem,
   }
 }
 
+template <bool static_entry>
 static void emit_indexed(grpc_chttp2_hpack_compressor* c, uint32_t elem_index,
                          framer_state* st) {
   GRPC_STATS_INC_HPACK_SEND_INDEXED();
-  uint32_t len = GRPC_CHTTP2_VARINT_LENGTH(elem_index, 1);
+  const uint32_t len =
+      static_entry ? 1 : GRPC_CHTTP2_VARINT_LENGTH(elem_index, 1);
```

Comment:
Could you explain the optimization here to me? It seems in the case of a static_entry, we could just call GRPC_CHTTP2_VARINT_LENGTH and it would return the desired length to us. Why is not entering GRPC_CHTTP2_VARINT_LENGTH and branching on static_entry an optimization?

My concern here is that even if it's a minor optimization, it may not justify the extra code complexity.

Reply:
This looks like a branch but isn't. Specifically, notice that we're branching on a template parameter, which means that it's resolved at compile time - so it's straight line code, which makes static element encoding about 2.5% faster (from the microbenchmarks in #20267.)

Reply:
Ah, I see. A followup question: why do the template parameter thing instead of having two methods? This seems to impair readability at the call site (e.g., a reader wouldn't know what the "false" or "true" template parameter means).

Reply:
From a functionality point of view it will be the same - either way the compiler will output two methods.

From a maintainability point of view: I have a preference for the template approach:
1. It reduces the chance that the two methods drift (in case someone updates one but not the other).
2. It makes it clear that they're fundamentally here to do the same thing.

While readability at the call site now has one extra step (jump to the method declaration to see what the template is) the alternative would be having to track more methods, which I think is harder in the long run.

Reply:
1. If we have two methods that have the common logic factored out into a helper function that they both call, doesn't that accomplish the same purpose of ensuring that the two methods don't drift?

2. I don't quite understand the point about tracking more methods. If the methods are clearly named, then the reader can understand the high level idea of what's going on at the call site. They won't have to jump to the method declaration just to figure that out, and I think this is valuable for readability.

Methods templated on a boolean seem to be as much of a code smell as boolean parameters.  From a readability perspective, I'd prefer the other approach.

Reply:
Since we have differing opinions here, I'll defer to the TL on whether to use boolean template parameter methods or make two separate methods: @markdroth

Reply:
I don't have a hard-and-fast rule about this, so it's as much a judgement call for me as for anyone else.  But as a practical matter, I don't tend to find a lot of cases where a bool template parameter really makes sense.  If the function is very small, then duplicating it isn't a big risk for their implementations diverging, because the two versions are not doing a whole lot.  And if the function is very large, then we probably don't want to templatize it, since that will bloat the generated code size; in that case, some refactoring is probably called for.

In this particular case, the function is so small that I think duplicating it is fine.  And please do add comments clarifying that the two versions do the same thing but are optimized for different cases.

Reply:
Sure, will go with that then.